### ISSUE TITLE
HTS Magnet - Step 2: Implement Ramping in the Kepco IOC

### ISSUE THREAD

**Issue by kjwoodsISIS** [ISISComputingGroup/IBEX#5661]
As a developer I would like to build up the new OPI and IOCs for the HTS magnet in discrete steps (the parent ticket is #5577).  Step 2 is to implement ramping in the Kepco IOC and to add the ramping fields to the OPI.

#### Acceptance Criteria
1. Using the Kepco IOC, I can
   1. I can set a ramp rate (units A/min) (the rate at which the current will be increased or decreased)
   1. The ramp rate must be a value that is greater than zero.
   1. On changing the current setpoint, the current is increased (or decreased) from its readback value to the value indicated by the setpoint.
   1. The current must not be changed at a rate that exceeds the ramp rate.
1. The current must be increased (or decreased) at `step_rate` steps per minute
   * the value of `step_rate` should be configurable.
1. It must be possible to configure a maximum ramp rate
   * ~suggestion: set the initial maximum ramp rate to 0.5 A/min~
   * manufacturer says maximum ramp rate is "around 40 A/min" (see [Additional Notes from Manufacturer](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/HTS_Magnet/HTS_Magnet_Answers_2020_09_03.pdf))
   * suggestion: set the initial maximum ramp rate to 35 A/min (maybe set it to 40A/min later).
1. It must be possible to configure a minimum ramp rate
   * the purpose of a minimum ramp rate is to prevent someone setting the ramp rate to less than 0.0 A/min.  The ramp rate must be greater than or equal to zero.
   * suggestion: set the initial minimum ramp rate to 0.0 A/min
1. The Ramp Control group box must contain:
   1. a ramp rate readback field (units A/min).
   1. a ramp rate setpoint field
1. It must be possible to configure a maximum current (a software limit)
   * ~suggestion: set the initial maximum current to 10.0 A~
   *   manufacturer says maximum current is 217A (see [Additional Notes from Manufacturer](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/HTS_Magnet/HTS_Magnet_Answers_2020_09_03.pdf))
1. It must be possible to configure a minimum current (a software limit)
   * suggestion: set the initial minimum current to 0.0 A

#### Notes
1. This ticket is a child of #5577.
1. Refer to the schematics to see how the OPI could be laid out:
   * [Main tab](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/HTS_Magnet/HTS_OPI_Schematics_v4_Normal_Tab.pdf)
1. A step-wise linear ramping scheme is suggested.  The IOC will need to perform a calculation along the following lines:
   1. The time (`T`) required to go from a starting current (`Is`) to a finishing current (`If`) is: `T = abs(If - Is) / ramp_rate`.
   1. The number of steps (`num_steps`) required to go from a starting current to a finishing current is: `num_steps = int(step_rate * T) + 1`.
   1. The change in current at each step (`step_size`) is: `step_size = (If - Is) / num_steps`.
   1. The interval (in seconds) between steps is `interval = 60 / step_rate`.
   1. Loop over num_steps incrementing (or decrementing if `step_size < 0`) the current every `interval` seconds.
1. A suggested ramping algorithm is illustrated in [Ramping.xslx](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/HTS_Magnet/Ramping.xlsx).
1. Tests should include tests that check correct ramping from:
   1. 0.0A to maximum current
   1. Maximum current to 0.0
   1. A non-zero current to a greater (non-zero) current.
   1. A non-zero current to a lower (non-zero) current.
1. During commissioning ask the scientists to propose maximum and minimum ramp rates.

**Comment by kjwoodsISIS** [comment]
More questions to scientists:
> From: Kevin
> To: Christy
> Sent: 27 August 2020 11:09
> Subject: Some more questions about the HTS Magnet
> 
> Hi Christy,
> 
> In planning the IOCs for the HTS magnet, some questions have cropped up.
> 
> The Kepco power supply does not have a built-in ramping feature.  Therefore, ramping of the 
> current has to be done in software.  How smooth does the ramping need to be?  Obviously, if 
> we set a ramp-rate of, say 1A/min, the current should not jump by 1A in a single step.  Rather, 
> the current should increase in a step-wise manner (e.g. by 0.25A every 15s).  So, one question 
> is:  how many steps are required to make the ramping sufficiently smooth?
> We also have to bear in mind that it takes time to send a command to the power supply and get 
> a response back (as much as 10s for the round trip), so there is a limit to the number of steps 
> we can make in a minute.
> 
> Setting the number of steps is one thing.  A related question is: what is the permitted maximum 
> step size?  For example, with a step frequency of 4 steps/minute and ramp-rate of 1A/min, the 
> step size is 0.25A.  However, if the ramp-rate was 4A/min, the step-size would be 1A.  
> [I appreciate my values in my examples may not be realistic – I chose them because they make 
> the arithmetic easy – but the point is valid.]  We will need to choose a maximum ramp rate that 
> limits the maximum step size.
> 
> Also, just setting a single value for the ramp-rate implies that the current increases (or 
> decreases) linearly (or, rather, in a step-wise linear fashion).  Is there a need to vary the 
> ramp-rate depending on the field (i.e. do we need a ramp profile, rather a simple ramp-rate)?
> 
> Finally, is there a maximum allowed current for the HTS magnet?  (We are assuming that there 
> is – we just don’t know what the value is).
> 
> I appreciate that you might not know the answers to these questions until the magnet actually 
> arrives on-site.  But from our perspective, the sooner we know the better. 
>

**Comment by kjwoodsISIS** [comment]
Answers to the above questions are stored in [ICP Discussions - HTS Magnet](http://www.facilities.rl.ac.uk/isis/computing/ICPdiscussions/HTS_Magnet/HTS_Magnet_Answers_2020_09_03.pdf).

**Comment by kjwoodsISIS** [comment]
The above answers will need discussion with the scientists.  In order to get a system up and running, we should implement a simple linear ramping mechanism.  Once we have reliable control of the system, we can consider (via another ticket) a more sophisticated ramping mechanism.

**Comment by JamesKingWork** [comment]
Ignore above PR it was attached to the wrong ticket

**Comment by DominicOram** [comment]
The algorithm used currently in IBEX for ramping in the Eurotherm is slightly different to the attached. Rather than use `step_size = (If - Is) / num_steps` we instead use `step_size = ramp_rate/step_rate`. If the `step_size` fits into `If - Is` with an integer amount these algorithms are the same. If it doesn't then the first algorithm modifies the `step_size` to be the same across the board whereas the second uses the supplied `step_size` until the last step, which is then the remainder. I believe the difference is very small.

@Kevin-Woods-Tessella Was this algorithm proposed by the instrument scientist? If so I'll check in that the new algorithm is ok.

**Comment by kjwoodsISIS** [comment]
No, the algorithm was my suggestion.  All the scientists want to do is set a ramp-rate.   The algorithm used in the Eurotherm will be fine, so long as the calculated `step-size` can't result in the algorithm exceeding the maximum `ramp_rate`.

**Comment by DominicOram** [comment]
Thanks @kjwoodsISIS. The OPI contains no way of setting:
* Whether the ramp is on/off
* What voltage the kepco is outputting
* The step rate

Are these values that unlikely to ever change and so can be covered by macros in the config. Note that if there is no ramp on/off button there is no easy way to abort a currently running ramp. Shall I ask the scientist for these details?

**Comment by kjwoodsISIS** [comment]
Thanks for highlighting these issues @DominicOram :
   1. Whether the ramp is on/off
      - I think this is an omission.  We should add a button to toggle the ramp on/off.
   1. What voltage the kepco is outputting
      - The HTS gets its power from two Kepcos connected in parallel (one slaved to the other).  The Kepco operates in current mode (meaning the voltage is fixed).  The voltage is set on the hardware.
   1. The step rate
      - The `step_rate` is intended to be a configurable value, not one that is set from the OPI.

**Comment by DominicOram** [comment]
PRs:
* https://github.com/ISISComputingGroup/EPICS-ReadASCII/pull/11
* https://github.com/ISISComputingGroup/EPICS-ioc/pull/545
* https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/380
* https://github.com/ISISComputingGroup/EPICS-kepco/pull/8
* https://github.com/ISISComputingGroup/ibex_gui/pull/1206

Improved documentation at https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/ISIS-modules-for-file-handling#readascii.

To test:
* Pull and build everything
* Confirm new IOC system tests pass
* Run with `-a` on the IOC system tests
* On the OPI turn ramping on and confirm that setting a set point performs a ramp
* Confirm with ramping off kepco behaviour is the same as before this ticket
* Confirm you can change the ramp rate and ramp increases/decreases
* Confirm you can start a KEPCO IOC with a MAX_CURRRENT macro and it will not accept currents higher than that
* Confirm you can start a KEPCO IOC with a STEP_NUMBER macro and the number of steps a ramp does changes (but the rate stays the same)

**Comment by RaiBishal** [comment]
Code looks good however, `CURRENT_MAX` and `STEPS_NUM` macros cannot be set from IBEX and can only be set via `set-common.cmd` file. 
The IOC test framework is failing at the new test because it is unable to find some PVs. I believe this is happening because it cannot the macros `CURRENT_MAX` and `STEPS_NUM` are not set.

**Comment by DominicOram** [comment]
Oops, sorry, I forgot to edit the `st.cmd`. I also got tripped up by the latest `IOC_test_framework` issue (see my post in dev changes). You need to make sure you're on master in EPICS top and to be on master in support\icpconfig and build icpconfig. Let me know if the tests still don't work then.
